crates/proof-of-sql/src/proof_primitive/dory/dynamic_dory_structure.rs

```diff
@@ -30,22 +30,62 @@
 //! ...
 //! ```
 
+/// Sets the data to indicate if no index exists.
+#[allow(dead_code)]
+const fn no_index() -> usize {
+    usize::MAX
+}
+
+/// Returns if the index exists.
+#[allow(dead_code)]
+pub(crate) const fn index_exists(index: usize) -> bool {
+    index != no_index()
+}
+
+/// Returns the full width of a row in the matrix.
+#[allow(dead_code)]
+pub(crate) const fn full_width_of_row(row: usize) -> usize {
+    ((2 * row + 4) / 3).next_power_of_two()
+}
+
 /// Returns the index that belongs in the first column in a particular row.
 ///
 /// Note: when row = 1, this correctly returns 0, even though no data belongs there.
 #[cfg(test)]
-pub(crate) fn row_start_index(row: usize) -> usize {
-    let width_of_row = ((2 * row + 4) / 3).next_power_of_two();
+pub(crate) const fn row_start_index(row: usize) -> usize {
+    let width_of_row = full_width_of_row(row);
     width_of_row * (row - width_of_row / 2)
 }
+
 /// Returns the (row, column) in the matrix where the data with the given index belongs.
-pub(crate) fn row_and_column_from_index(index: usize) -> (usize, usize) {
+pub(crate) const fn row_and_column_from_index(index: usize) -> (usize, usize) {
     let width_of_row = 1 << (((2 * index + 1).ilog2() + 1) / 2);
     let row = index / width_of_row + width_of_row / 2;
     let column = index % width_of_row;
     (row, column)
 }
 
+/// Returns the index of data where the (row, column) belongs.
+#[allow(dead_code)]
+pub(crate) const fn index_from_row_and_column(row: usize, column: usize) -> usize {
+    let width_of_row = full_width_of_row(row);
+
+    // Return no_index if the (row, column) is not in dynamic Dory structure.
+    if column >= width_of_row || (row == 1 && column == 0) {
+        return no_index();
+    }
+
+    width_of_row * (row - width_of_row / 2) + column
+}
```

Comment:
This is a textbook example for why the `Option` type exists in Rust.
```suggestion
pub(crate) const fn index_from_row_and_column(row: usize, column: usize) -> Option<usize> {
    let width_of_row = full_width_of_row(row);

    // Return no_index if the (row, column) is not in dynamic Dory structure.
    if column >= width_of_row || (row == 1 && column == 0) {
        return None;
    } else {
        Some(width_of_row * (row - width_of_row / 2) + column)
    }
}
```

Or, equivalently;

```rust
pub(crate) const fn index_from_row_and_column(row: usize, column: usize) -> Option<usize> {
    let width_of_row = full_width_of_row(row);
    (column < width_of_row && (row, column) != (1, 0))
        .then_some(width_of_row * (row - width_of_row / 2) + column)
}
```

Reply:
Updated, thank you!